core/drivers/crypto/se050/core/ecc.c

```diff
@@ -0,0 +1,627 @@
+	ret = set_binary_data(public_key->x, x1_len, &key.x, &key.x_len);
+	if (ret != TEE_SUCCESS)
+		return ret;
+
+	ret = set_binary_data(public_key->y, y1_len, &key.y, &key.y_len);
+	if (ret != TEE_SUCCESS) {
+		free(key.x);
+		return ret;
+	}
+
+	st = se050_ecc_gen_shared_secret(&se050_session->s_ctx, kid, &key,
+					 secret, secret_len);
+	free(key.x);
+	free(key.y);
+	if (st != kStatus_SSS_Success)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result sign(uint32_t algo, struct ecc_keypair *key,
+		       const uint8_t *msg, size_t msg_len,
+		       uint8_t *sig, size_t *sig_len)
+{
+	sss_status_t st = kStatus_SSS_Fail;
+	sss_se05x_asymmetric_t ctx = { };
+	sss_se05x_object_t kobject = { };
+	TEE_Result res = TEE_SUCCESS;
+	size_t key_bytes = 0;
+	size_t key_bits = 0;
+
+	res = ecc_get_key_size(key->curve, algo, &key_bytes, &key_bits);
+	if (res != TEE_SUCCESS)
+		goto exit;
+
+	/* se050 exports DER format */
+	if (*sig_len < (2 * key_bytes + DER_SIGNATURE_SZ)) {
+		*sig_len = 2 * key_bytes + DER_SIGNATURE_SZ;
+		res = TEE_ERROR_SHORT_BUFFER;
+		goto exit;
+	}
+
+	res = ecc_get_msg_size(algo_tee2se050(algo), &msg_len);
+	if (res != TEE_SUCCESS)
+		goto exit;
+
+	res = se050_inject_keypair(&kobject, key, key_bytes);
+	if (res != TEE_SUCCESS)
+		goto exit;
+
+	st = sss_se05x_asymmetric_context_init(&ctx, se050_session, &kobject,
+					       algo_tee2se050(algo),
+					       kMode_SSS_Sign);
+	if (st != kStatus_SSS_Success) {
+		sss_se05x_key_store_erase_key(se050_kstore, &kobject);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	st = sss_se05x_asymmetric_sign_digest(&ctx, (uint8_t *)msg, msg_len,
+					      sig, sig_len);
+	if (st != kStatus_SSS_Success) {
+		EMSG("curve: 0x%x", key->curve);
+		res = TEE_ERROR_BAD_PARAMETERS;
+		goto exit;
+	}
+
+	sss_se05x_signature_der2bin(sig, sig_len);
+exit:
+	if (!se050_ecc_keypair_from_nvm(key))
+		sss_se05x_key_store_erase_key(se050_kstore, &kobject);
+
+	sss_se05x_asymmetric_context_free(&ctx);
+
+	return res;
+}
+
+static TEE_Result verify(uint32_t algo, struct ecc_public_key *key,
+			 const uint8_t *msg, size_t msg_len,
+			 const uint8_t *sig, size_t sig_len)
+{
+	sss_status_t st = kStatus_SSS_Fail;
+	sss_se05x_asymmetric_t ctx = { };
+	sss_se05x_object_t kobject = { };
+	TEE_Result res = TEE_SUCCESS;
+	uint8_t signature[128];
+	size_t signature_len = sizeof(signature);
+	size_t key_bytes = 0;
+	size_t key_bits = 0;
+
+	res = ecc_get_key_size(key->curve, algo, &key_bytes, &key_bits);
+	if (res != TEE_SUCCESS)
+		goto exit;
+
+	res = ecc_get_msg_size(algo_tee2se050(algo), &msg_len);
+	if (res != TEE_SUCCESS)
+		goto exit;
+
+	res = se050_inject_public_key(&kobject, key, key_bytes);
+	if (res != TEE_SUCCESS)
+		goto exit;
+
+	st = sss_se05x_asymmetric_context_init(&ctx, se050_session, &kobject,
+					       algo_tee2se050(algo),
+					       kMode_SSS_Verify);
+	if (st != kStatus_SSS_Success) {
+		sss_se05x_key_store_erase_key(se050_kstore, &kobject);
```

Comment:
Above you're using the `exit:` label for cleanup, but not here. What's different?

Reply:
fixing it.